virtcontainers/qemu.go

```diff
@@ -584,6 +588,37 @@ func (q *qemu) createSandbox(ctx context.Context, id string, networkNS NetworkNa
 		return err
 	}
 
+	/*
+		Solve MMIO space over 4G for PCIe hotplug:
+		Issue:
+		Some large bar space PCIe devices (for example: Nvidia Tesla P100) will be not able to assigned BAR space.
+		A similar error message will appear in dmesg,
+			[   14.807071] pci 0000:00:06.0: BAR 1: no space for [mem size 0x400000000 64bit pref]
+			[   14.807891] pci 0000:00:06.0: BAR 1: failed to assign [mem size 0x400000000 64bit pref]
+		Solution: (https://github.com/qemu/qemu/blob/master/docs/pcie.txt)
+		- Plugging a PCIe device into a PCI slot will hide the Extended Configuration Space thus is also not recommended.
+		- The PCIe features are available only when using the 'q35'
+		- PCIe devices should be plugged only into PCI Express Root Ports and PCI Express Downstream ports.
+		  Prefer grouping PCI Express Root Ports into multi-function devices to keep a simple flat hierarchy that is
+		  enough for most scenarios.
+		- The pcie.0 bus do not support hot-plug, so any devices plugged into pcie.0 cannot be hot-plugged/hot-unplugged,
+		  include the PCI Express Root Ports.
+		So we need to add PCI Express Root Port to support hot-plug PCIe devices in q35.
+		Note:
+		- Only vfio devices will be hot-plugged into PCI Express Root Ports.
+	*/
+	if hypervisorConfig.HypervisorMachineType == QemuQ35 {
+		for i := uint(0); i < hypervisorConfig.PCIeRootPort; i++ {
+			pciePortRootDev := config.PCIeRootPortDev{
+				ID:      fmt.Sprintf("%s%d", pcieRootPortPrefix, i),
+				Bus:     defaultBridgeBus,
+				Chassis: "0",
+				Slot:    strconv.FormatUint(uint64(i), 10),
+			}
```

Comment:
move this to qemu_amd64.go (`appendPCIeRootPortDevice`), since this is x86 specific

Reply:
OK